app/tests/reader_studies_tests/test_api.py

```diff
@@ -797,9 +802,11 @@ def test_answer_creator_is_reader(client):
         (Question.AnswerType.MULTIPLE_POLYGONS, None, 201),
         (Question.AnswerType.ANGLE, None, 201),
         (Question.AnswerType.MULTIPLE_ANGLES, None, 201),
-        (Question.AnswerType.CHOICE, None, 400),
+        (Question.AnswerType.CHOICE, None, 201),
```

Comment:
Hmm, now we're saying that non-answers for Choice and Number are now `None` rather than `""`. Has something changed in CIRRUS? Do all widgets now return these answers rather than `""`? It must have worked before surely? We would need to check the database and existing values are consistent.

Reply:
I assumed non-answers representation based on the snippets below and the errors that the validators are reporting about None types. I'll double-check what CIRRUS outputs atm.


https://github.com/comic/grand-challenge.org/blob/8e1e08b5d6d2a94752c2b243f98105a128449a3d/app/grandchallenge/reader_studies/models.py#L1436-L1449

Reply:
Widgets haven't changed AFAIK. The current output of CIRRUS with no answers:

```
AnswerType.SINGLE_LINE_TEXT ""
AnswerType.MULTI_LINE_TEXT ""
AnswerType.TEXT ""
AnswerType.BOOL False
AnswerType.NUMBER None
AnswerType.CHOICE None
AnswerType.MULTIPLE_CHOICE []
AnswerType.POINT None
AnswerType.MULTIPLE_POINTS None
AnswerType.DISTANCE_MEASUREMENT None
AnswerType.MULTIPLE_DISTANCE_MEASUREMENTS None
AnswerType.BOUNDING_BOX_2D None
AnswerType.MULTIPLE_2D_BOUNDING_BOXES None
AnswerType.POLYGON None
AnswerType.MULTIPLE_POLYGONS None
AnswerType.LINE None
AnswerType.MULTIPLE_LINES None
AnswerType.MASK None
AnswerType.ANGLE None
AnswerType.MULTIPLE_ANGLES None
AnswerType.ELLIPSE None
AnswerType.MULTIPLE_ELLIPSES None
```

Reply:
`CHOICE` and `NUMBER` all annotations have always (or at least for a while) had non-answer as `None`. I wasn't too sure about the `NUMBER` but I checked what CIRRUS (pre-widget split) does and v6.0.0 of CIRRUS already produces `None` for `NUMBER`.

We do sort of break with the Widget (HTML) wise in that the number inputs always generate a `""` when left empty. But then again, that is always a string... so even a basic interpretation should put it to `None`/`Null` before processing.

So sorry, I cannot follow where you see the hiccup.

Reply:
OK, strange, then I have no idea what is/was going on with these tests. We indeed only have values for `None` for those types:

```python
>>> Answer.objects.filter(answer__isnull=True, question__answer_type=AnswerType.CHOICE).count()
124
>>> Answer.objects.filter(answer="", question__answer_type=AnswerType.CHOICE).count()
0
>>> Answer.objects.filter(answer__isnull=True, question__answer_type=AnswerType.NUMBER).count()
4536
>>> Answer.objects.filter(answer="", question__answer_type=AnswerType.NUMBER).count()
0
```

The tests here would suggest that they wouldn't be accepted. Were they changed at an earlier point?

Reply:
The test for the CHOICE field were added here, an apparently was never nullable? https://github.com/comic/grand-challenge.org/commit/377d1496fca0c04dc90cbc586dcde7aff91ee055 This doesn't make sense to me, what am I missing?

Reply:
The first null NUMBER or CHOICE answers were created last year:

```python
>>> Answer.objects.filter(answer__isnull=True, question__answer_type=AnswerType.NUMBER).order_by("created").first().created
datetime.datetime(2022, 3, 23, 14, 21, 53, 800920, tzinfo=datetime.timezone.utc)
>>> Answer.objects.filter(answer__isnull=True, question__answer_type=AnswerType.CHOICE).order_by("created").first().created
datetime.datetime(2021, 10, 5, 14, 13, 34, 282831, tzinfo=datetime.timezone.utc)
```

How? Is this test actually testing what it is supposed to?

Reply:
I think we should have a separate test with that checks that blank answers are accepted regardless of the required status:

```python
@pytest.mark.parametrize(
  "answer_type,answer",
  (
AnswerType.SINGLE_LINE_TEXT ""
AnswerType.MULTI_LINE_TEXT ""
AnswerType.TEXT ""
AnswerType.BOOL False
AnswerType.NUMBER None
AnswerType.CHOICE None
AnswerType.MULTIPLE_CHOICE []
AnswerType.POINT None
AnswerType.MULTIPLE_POINTS None
AnswerType.DISTANCE_MEASUREMENT None
AnswerType.MULTIPLE_DISTANCE_MEASUREMENTS None
AnswerType.BOUNDING_BOX_2D None
AnswerType.MULTIPLE_2D_BOUNDING_BOXES None
AnswerType.POLYGON None
AnswerType.MULTIPLE_POLYGONS None
AnswerType.LINE None
AnswerType.MULTIPLE_LINES None
AnswerType.MASK None
AnswerType.ANGLE None
AnswerType.MULTIPLE_ANGLES None
AnswerType.ELLIPSE None
AnswerType.MULTIPLE_ELLIPSES None
))
@pytest.mark.parametrize(
  "is_required",
  (False, True, ),
)
def test_blank_answers(answer_type, answer, is_required):
  ...
  assert response.status_code == 201
```

But obviously valid python.

Reply:
I have added your suggested test. TIL you can chain `parametrize()`. This focuses on the required answers and subtypes. 

However, I don't think we can/should revert the `test_answer_is_correct_type`. For that specific test, the `None` answers for annotations would stumble over the new logic for required questions and answers. Setting the test question to not required allows us to test the typing better: both empty answers and regular answers are testable for validity.

I re-arranged the cases a bit to bundle the answer_types together. Imho this makes this a bit more readable.